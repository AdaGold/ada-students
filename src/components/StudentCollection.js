// src/components/StudentCollection.js
import React from 'react';
import PropTypes from 'prop-types';

import Student from './Student';

const StudentCollection = (props) => {

  const studentComponents = props.students.map((student, i) => {
    return (
      <li key={i}>
        <Student
          fullName={student.fullName}
          email={student.email}
          present={student.present}
          id={student.id}
          onUpdateStudent={props.onUpdateStudent}
        />
      </li>
    );
  });

  return (
    <ul className="student-collection">
      {studentComponents}
    </ul>
  );
};

StudentCollection.propTypes = {
  students: PropTypes.arrayOf(PropTypes.shape(
    {
      fullName: PropTypes.string.isRequired,
      email: PropTypes.string,
      present: PropTypes.bool,
      id: PropTypes.number.isRequired,
    },
  )),
  onUpdateStudent: PropTypes.func.isRequired,
}



export default StudentCollection;
