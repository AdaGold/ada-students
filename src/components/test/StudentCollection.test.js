import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react'
import StudentCollection from '../StudentCollection';

describe('StudentCollection', () => {

  afterEach(cleanup);

  const students = [
    {
      id: 1,
      fullName: "Serena Williams",
      email: "serena@williams.com",
      present: true,
    },
    {
      id: 2,
      fullName: "Vemus Williams",
      email: "venus@williams.com",
      present: false,
    },
  ]

  test('will render properly', () => {
    // Arrange-Act
    const { asFragment } = render(
      <StudentCollection
        students={students}
        onUpdateStudent={() => { }}
      />
    );

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  test('will the proper number of stud', () => {
    // Arrange-Act
    const { getAllByText } = render(
      <StudentCollection
        students={students}
        onUpdateStudent={() => { }}
      />
    );

    // Assert
    expect(getAllByText(/Class: C13/i).length).toEqual(2);
  });

  test('will forward on the onUpdateStudent event handler', () => {
    // Arrange-Act
    const onUpdateStudentCallbackFunction = jest.fn(() => { });
    const { getAllByText } = render(
      <StudentCollection
        students={students}
        onUpdateStudent={onUpdateStudentCallbackFunction}
      />
    );

    // get the first mark present button
    const button = getAllByText(/Mark Present/i)[0];
    // Click it
    fireEvent.click(button);

    // Assert that the callback function is called
    expect(onUpdateStudentCallbackFunction).toHaveBeenCalled();
  });
});