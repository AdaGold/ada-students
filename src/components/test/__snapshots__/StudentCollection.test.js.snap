// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentCollection will render properly 1`] = `
<DocumentFragment>
  <ul
    class="student-collection"
  >
    <li>
      <div
        class="student"
      >
        <h3
          class="present"
        >
          Serena Williams
        </h3>
        <input
          value="Serena Williams"
        />
        <ul>
          <li>
            Class: C13
          </li>
          <li>
            Birthday: no birthdate on file
          </li>
          <li>
            Email: serena@williams.com
          </li>
        </ul>
        <button>
          Mark Absent
        </button>
      </div>
    </li>
    <li>
      <div
        class="student"
      >
        <h3
          class="absent"
        >
          Vemus Williams
        </h3>
        <input
          value="Vemus Williams"
        />
        <ul>
          <li>
            Class: C13
          </li>
          <li>
            Birthday: no birthdate on file
          </li>
          <li>
            Email: venus@williams.com
          </li>
        </ul>
        <button>
          Mark Present
        </button>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
